### DISCUSSION TITLE
[Proposal] Local functions for properties that are shared between the getter and setter

### DISCUSSION THREAD

**Discussion by AustinBryan** [dotnet/csharplang#8755] [Language Ideas]
Currently, we can have local functions for a getter or a setter, like so:

```csharp
int Foo
{
    get 
    {
        return _foo + Bar();

        int Bar() { ... }
    }
}
```

However, I don't find the need for this that common. What I do think is more useful is being able to define a function that is scoped to the property such that *both* the getter and setter can use it, as sometimes they need the same tools to work with.

### Example:

```csharp
int?[,] array;

this[Coord coord]
{
    get => (coord.X >= 0 && coord.X < array.GetLength(0) && coord.Y >= 0 && coord.Y < array.GetLength(1)) ? array[coord.X, coord.Y] : null;
    set
    {
    	if ((coord.X >= 0 && coord.X < array.GetLength(0) && coord.Y >= 0 && coord.Y < array.GetLength(1)))
            array[coord.X, coord.Y] = value;
    }
}
```

This isn't easy to read and requires writing the same thing twice. The current way to do this would be to extract the method, but then it's visible to the whole class, which isn't neccessary, seeing as it's only used inside the property. 

### Proposed Solution:

```csharp
int?[,] array;

this[Coord coord]
{
    get => CoordInRange() ? array[coord.X, coord.Y] : null;
    set
    {
    	if (CoordInRange()) array[coord.X, coord.Y] = value;
    }
    
    bool CoordInRange() => coord.X >= 0 && coord.X < array.GetLength(0) && coord.Y >= 0 && coord.Y < array.GetLength(1);
}
```

To me, this is much cleaner. They both can use `CoordInRange`, the code is much easier to read, and I'm not repeating any code, nor am I exposing the function that's only used once to the entire class.

**Comment by HaloFour** [comment]
This seems to be in the same vein as #133.  If property-scoped fields are considered I don't see why they also couldn't include property-scoped functions.

**Comment by CyrusNajmabadi** [comment]
And then property scoped nested types!

**Comment by Neme12** [comment]
If we have property scoped fields and property scoped methods... what about property scoped properties? :smile:

**Comment by HaloFour** [comment]
@Neme12   Yo dawg, I heard you like properties ...
